packages/backend/src/services/ProjectService/ProjectService.ts

```diff
@@ -1081,48 +1089,65 @@ export class ProjectService extends BaseService {
             );
         }
 
-        // If requireUserCredentials is true, we need to override the existing credentials with the user credentials
-        // even if we use organization credentials
-        if (credentials.requireUserCredentials) {
-            if (!isRegisteredUser) {
-                throw new ForbiddenError(
-                    'Embedded users cannot use personal warehouse credentials',
-                );
-            }
+        // Check if user has their own credentials for this project's warehouse type
+        // Only fetch user credentials when:
+        // 1. requireUserCredentials is enabled (user credentials are mandatory)
+        // 2. Databricks warehouse (supports optional user OAuth credentials)
+        const shouldFetchUserCredentials =
+            credentials.requireUserCredentials ||
+            credentials.type === WarehouseTypes.DATABRICKS;
+
+        if (isRegisteredUser) {
+            // Fetch user credentials only when needed (for performance)
+            const userWarehouseCredentials = shouldFetchUserCredentials
+                ? await this.userWarehouseCredentialsModel.findForProjectWithSecrets(
+                      projectUuid,
+                      userId,
+                      credentials.type,
+                  )
+                : undefined;
 
-            const userWarehouseCredentials =
-                await this.userWarehouseCredentialsModel.findForProjectWithSecrets(
-                    projectUuid,
+            if (userWarehouseCredentials !== undefined) {
```

Comment:
Why an exact check for `undefined` rather than a simple truthy `(userWarehouseCredentials)` like we do below on line 1133 with `else if (credentials.requireUserCredentials)`? Maybe just me, but for JS/TS, I start looking around to see if we're handling other falsey-ish cases vs. other languages where the empty state is stricter. Not a blocker at all, just curious.